dids/did_dht_test.go

```diff
@@ -0,0 +1,170 @@
+package dids
+
+import (
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"testing"
+
+	"github.com/alecthomas/assert/v2"
+	"golang.org/x/net/dns/dnsmessage"
+)
+
+type bep44TXTResourceOpt func() dnsmessage.Resource
+
+func WithDNSRecord(name, body string) bep44TXTResourceOpt {
+	return func() dnsmessage.Resource {
+		return dnsmessage.Resource{
+			Header: dnsmessage.ResourceHeader{
+				Name: dnsmessage.MustNewName(name),
+				Type: dnsmessage.TypeTXT,
+				TTL:  7200,
+			},
+			Body: &dnsmessage.TXTResource{
+				TXT: []string{
+					body,
+				},
+			},
+		}
+	}
+}
+func makeDNSMessage(answersOpt ...bep44TXTResourceOpt) dnsmessage.Message {
+
+	answers := []dnsmessage.Resource{}
+	for _, a := range answersOpt {
+		answers = append(answers, a())
+	}
+
+	msg := dnsmessage.Message{
+		Header:  dnsmessage.Header{Response: true, Authoritative: true},
+		Answers: answers,
+	}
+
+	return msg
+}
+
+func TestDHTResolve(t *testing.T) {
+	// vector taken from https://github.com/TBD54566975/web5-js/blob/dids-new-crypto/packages/crypto/tests/fixtures/test-vectors/secp256k1/bytes-to-public-key.json
+	publicKeyHexSecp256k1 := "0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8"
+	pubKeyBytesSecp256k1, err := hex.DecodeString(publicKeyHexSecp256k1)
+	assert.NoError(t, err)
+	base64EncodedSecp256k := base64.RawURLEncoding.EncodeToString(pubKeyBytesSecp256k1)
+
+	tests := map[string]struct {
+		didURI        string
+		msg           dnsmessage.Message
+		expectedError error
+		assertResult  func(t *testing.T, d *Document)
+	}{
+		"did with valid key and no service": {
+			didURI: "did:dht:cwxob5rbhhu3z9x3gfqy6cthqgm6ngrh4k8s615n7pw11czoq4fy",
+			msg: makeDNSMessage(
+				WithDNSRecord("_did.", "vm=k0;auth=k0;asm=k0;inv=k0;del=k0"),
+				WithDNSRecord("_k0._did.", "id=0;t=0;k=YCcHYL2sYNPDlKaALcEmll2HHyT968M4UWbr-9CFGWE"),
+			),
+			expectedError: nil,
+			assertResult: func(t *testing.T, d *Document) {
+				assert.False(t, d == nil, "Expected non nil document")
+				assert.NotZero[string](t, d.ID, "Expected DID Document ID to be initialized")
+				assert.NotZero[[]VerificationMethod](t, d.VerificationMethod, "Expected at least 1 verification method")
+			},
+		},
+		"did with multiple valid keys and no service - out of order verification methods": {
+			didURI: "did:dht:cwxob5rbhhu3z9x3gfqy6cthqgm6ngrh4k8s615n7pw11czoq4fy",
+			msg: makeDNSMessage(
+				WithDNSRecord("_did.", "vm=k0,k1,k2;auth=k0;asm=k1;inv=k2;del=k0"),
+				WithDNSRecord("_k0._did.", "id=0;t=0;k=YCcHYL2sYNPDlKaALcEmll2HHyT968M4UWbr-9CFGWE"),
+				WithDNSRecord("_k2._did.", fmt.Sprintf("id=2;t=1;k=%s", base64EncodedSecp256k)),
+				WithDNSRecord("_k1._did.", fmt.Sprintf("id=1;t=2;k=%s", base64EncodedSecp256k)),
+			),
+			expectedError: nil,
+			assertResult: func(t *testing.T, d *Document) {
+				assert.False(t, d == nil, "Expected non nil document")
```

Comment:
You can use `assert.NotZero(t, d, "...")` if you prefer

Reply:
Is that also checking nil or just the "zero value"? I think I'm ok leaving it like this for now, it's more explicit

Reply:
nil is the zero value for d, so yes.